pkg/fqdn/cache.go

```diff
@@ -0,0 +1,191 @@
+// unique IPs are returned in separate lookups.
+// Redundant or expired entries are removed on insert.
+// Lookups check for expired entries.
+type DNSCache struct {
+	lock.RWMutex
+
+	// forward DNS lookups name -> IPEntries
+	// IPEntries maps IP -> entry that provides it. An entry may provide multiple IPs.
+	forward map[string]ipEntries
+}
+
+// NewDNSCache returns an initialized DNSCache
+func NewDNSCache() *DNSCache {
+	c := &DNSCache{
+		forward: make(map[string]ipEntries),
+	}
+
+	return c
+}
+
+// Update inserts a new entry into the cache.
+// After insertion cache entries for name are expired and redundant entries
+// evicted. This is O(number of new IPs) for eviction, and O(number of IPs for
+// name) for expiration.
+func (c *DNSCache) Update(lookupTime time.Time, name string, ips []net.IP, ttl int) {
+	entry := &cacheEntry{
+		Name:           name,
+		LookupTime:     lookupTime,
+		ExpirationTime: lookupTime.Add(time.Duration(ttl) * time.Second),
+		TTL:            ttl,
+		IPs:            ips,
+	}
+
+	c.Lock()
+	defer c.Unlock()
+
+	entries, exists := c.forward[name]
+	if !exists {
+		entries = make(map[string]*cacheEntry)
+		c.forward[name] = entries
+	}
+	c.updateWithEntryIPs(entries, entry)
+	// When lookupTime is much earlier than time.Now(), we may not expire all
+	// entries that should be expired, leaving more work for .Lookup.
+	c.removeExpired(entries, time.Now())
+}
+
+// Lookup returns a set of unique IPs that are currently unexpired for name, if
+// any exist. An empty list indicates no valid records exist. The IPs are
+// returned sorted.
+func (c *DNSCache) Lookup(name string) (ips []net.IP) {
+	return c.lookupByTime(time.Now(), name)
+}
+
+// lookupByTime takes a timestamp for expiration comparisions, and is only
+// intended for testing.
+func (c *DNSCache) lookupByTime(now time.Time, name string) (ips []net.IP) {
+	c.RLock()
+	defer c.RUnlock()
+
+	entries, found := c.forward[name]
+	if !found {
+		return nil
+	}
+
+	return entries.getIPs(now)
+}
+
+// updateWithEntry adds a mapping for every IP found in `entry` to `ipEntries`
+// (which maps IP -> cacheEntry). It will replace existing IP->old mappings in
+// `entries` if the current entry expires sooner (or has already expired).
+// This needs a write lock
+func (c *DNSCache) updateWithEntryIPs(entries ipEntries, entry *cacheEntry) {
+	for _, ip := range entry.IPs {
+		ipStr := ip.String()
+		old, exists := entries[ipStr]
+		if old == nil || !exists || old.isExpiredBy(entry.ExpirationTime) {
+			entries[ipStr] = entry
+		}
+	}
+}
+
+// removeExpired removes expired (or nil) cacheEntry pointers from entries, an
+// ipEntries for a specific name.
+// This needs a write lock
+func (c *DNSCache) removeExpired(entries ipEntries, now time.Time) {
+	for ip, entry := range entries {
+		if entry == nil || entry.isExpiredBy(now) {
+			delete(entries, ip)
+		}
+	}
+}
+
+// keepUniqueIPs returns part of ips that includes only 1 entry per unique IP,
```

Comment:
This is extremely nitpicky (and doesn't really matter) - why not describe this in terms of bags (multisets) / sets? I.e., "transforms the provided multiset of IPs into a single set sorted lexicographically sorted..."

Reply:
"transform" is a good word to use here. I really wanted to convey that it was a mutation not a copy. I've basically copy pasted your sentence :P